porting of imau climate matrix

fretamal-ramirez opened the pull request:
Add code to run the climate matrix with warm-cold snapshots and CO2 forcing. It includes subroutines to allocate and read the climate snapshots. I tried to keep it as closer as possible to the original code but maybe it could be better to move some subroutines of the code to something like climate_utilities.f90. Let me know what do you think :)

TijnBerends commented on src/climate/climate_realistic.f90:
```diff
@@ -257,45 +269,46 @@ SUBROUTINE initialise_insolation_forcing( climate, mesh)
+      ALLOCATE(snapshot%ins_lat            (   snapshot%ins_nlat))
+      ALLOCATE(snapshot%ins_Q_TOA0         (mesh%vi1:mesh%vi2,12))
+      ALLOCATE(snapshot%ins_Q_TOA1         (mesh%vi1:mesh%vi2,12))
+      !ALLOCATE(snapshot%lambda             (mesh%vi1:mesh%vi2)) ! initialised in climate matrix routine
```
This points to an underlying problem. NO need to fix it right now, but eventually it would be nicer to clean this up, and place the calculation of the lapse rate for a snapshot within the initialise_snapshot routine.

TijnBerends commented on src/climate/climate_realistic.f90:
```diff
@@ -257,45 +269,46 @@ SUBROUTINE initialise_insolation_forcing( climate, mesh)
-        climate%snapshot%ins_t0 = closest_t0
-        call read_field_from_file_0D( C%filename_insolation, field_name_options_time, climate%snapshot%ins_t1, time_to_read = C%start_time_of_run+1000._dp)
+        snapshot%ins_t0 = closest_t0
+        call read_field_from_file_0D( C%filename_insolation, field_name_options_time, snapshot%ins_t1, time_to_read = C%start_time_of_run+1000._dp)
```
This seems to assume that the NetCDF file containing the insolation data has a temporal resolution of 1000 years. It would be better to make this more flexible.

TijnBerends commented on src/climate/climate_realistic.f90:
```diff
@@ -426,6 +440,166 @@ SUBROUTINE update_insolation_timeframes_from_file( climate, time, mesh)
+      IF (par%primary)  WRITE(*,"(A)") ' Initialising CO2 record '
+
+    ! Allocate shared memory to take the data
+    allocate( forcing%CO2_time(   C%CO2_record_length))
```
The "CO2_record_length" variable dates back to the time when we read CO2 data from a text file. With NetCDF data this is no longer needed, just read the size of the data from the NetCDF file.

TijnBerends commented on src/main/UFEMISM_main_model.f90:
```diff
@@ -484,6 +484,7 @@ SUBROUTINE initialise_model_region( region, region_name, forcing, start_time_of_
        lambda_M = region%mesh%lambda_M, phi_M = region%mesh%phi_M, beta_stereo = region%mesh%beta_stereo)
 
     ! Set up a regional frocing data type so the regional models can run asynchronously
+    ! this causes trouble during running, it will initialise regional_forcing instead of forcing, so variables will not be allocated later on for the run
```
Again, this indicates an underlying problem. 'forcing', being of 'type_global_forcing', should, as the name suggests, containing global forcing data; either globally uniform scalars that changes over time (e.g. benthic d18O, CO2), or global, mesh-independent fields (e.g. insolation as a function of latitude). To translate this data to something UFEMISM can use, there should also be a 'type_regional_forcing', which should be a part of 'type_model_region', i.e. region%forcing. Within run_model_region, there should then be a call to 'map_global_forcing_to_model_region', which takes 'forcing' as input, and maps it to region%mesh at that model time.

TijnBerends reviewed (changes requested):
There's a few issues here and there. Some of these are problems that already existed in UFE1.x, where the climate matrix code is several years old and back then I wasn't a very good programmer yet... Nevertheless, we'll apply the boy-scout rule: leave the code nicer than you found it. The problem with the global forcing data being altered within the regional model is one that I really think should be solved, because this inconsistency is pretty much guaranteed to cause trouble when you start doing simulations with more than one ice sheet.

fretamal-ramirez commented on src/climate/climate_matrix.f90:
```diff
@@ -0,0 +1,1279 @@
+
+! here appear some checks if wind is included or not in the data
+! after that cames the snapshot ocean and ice mask in dev branch, not in UFE1.x
+! need to check what it is? maybe is good to implement it. LINE 1215
```
Noo, it is in the dev branch of IMAU-ICE. It calls the subroutine initialise_snapshot_mask for each snapshot

fretamal-ramirez commented:
Hey, I have done most of the changes that you asked for

The following routines were moved to climate/matrix/climate_matrix_utilities.f90
- allocate_climate_snapshot
- read_climate_snapshot
- adapt_precip_CC
- adapt_precip_Roe
- precipitation_model_Roe
- rotate_wind_to_model_mesh
- get_insolation_at_time
- update_insolation_timeframes_from_file

Split the routines: 
- run_climate_model_matrix_temperature and run_climate_model_matrix_precipitation that now use weighting_fields_matrix_temperature and weighting_fields_matrix_precipitation, respectively.
- initialise_matrix_calc_spatially_variable_lapserate use calc_regional_avrg_temperature_outside_icesheet and calc_regional_avrg_temperature_outside_icesheet

These subroutines are still on climate_matrix.f90, but I can move them to climate_matrix_utilities.f90

I also moved the routines: initialise_CO2_record and update_CO2_at_model_time to global_forcings_main.f90, they are now called in initialise_global_forcings and update_global forcings.

I have a question regarding the use of a glacial index. Now I have only added the option C%climate_matrix_switch_glacial_index_precip that is used in the precipitation component; however, we can change it to C%climate_matrix_switch_glacial_index to call it also in the temperature component with something like UFE1.x with 

if (C%climate_matrix_switch_glacial_index) then
  w_tot = w_CO2
end if

Let me know what you think :)

TijnBerends commented on src/climate/climate_main.f90:
```diff
@@ -366,6 +387,9 @@ SUBROUTINE remap_climate_model( mesh_old, mesh_new, climate, region_name)
     TYPE(type_mesh),                        INTENT(IN)    :: mesh_new
     TYPE(type_climate_model),               INTENT(INOUT) :: climate
     CHARACTER(LEN=3),                       INTENT(IN)    :: region_name
+    type(type_grid), optional,                    intent(in)    :: grid
```
There's no need for these three arguments to be optional. Calling remap_climate_model without providing them implies that you already know that you're running a climate model other than the climate matrix, while the point of wrappers like "run_climate_model", "remap_climate_model", etc. is that they are called the same way regardles of which climate model you're actually running.

TijnBerends commented on src/climate/climate_matrix.f90:
```diff
@@ -0,0 +1,974 @@
+    call get_insolation_at_time( mesh, time, climate%snapshot)
+
+    ! moved to global_forcings_main
+    !CALL update_CO2_at_model_time( time, forcing)
```
Nice!

TijnBerends commented on src/climate/climate_matrix.f90:
```diff
@@ -0,0 +1,974 @@
+    ! Local variables:
+    character(LEN=256), parameter                      :: routine_name = 'weighting_fields_matrix_temperature'
+    integer                                            :: vi ,m
+    real(dp), DIMENSION(:    ), allocatable            :: w_ins,  w_ins_smooth,  w_ice
```
You don't need to change this right now, but there's no need for these arrays to be allocatable; as soon as we enter this subroutine, we already know what size they should be (namely mesh%vi1:mesh%vi2), so just real(dp), dimension(mesh%vi1:mesh%vi2) :: w_ins, etc. is fine.

TijnBerends commented on src/climate/matrix/climate_matrix_utilities.f90:
```diff
@@ -0,0 +1,442 @@
+  public :: read_climate_snapshot
+  public :: adapt_precip_CC
+  public :: adapt_precip_Roe
+  !public :: rotate_wind_to_model_mesh ! I don't think is needed...
```
If I remember correctly, the ERA5 data we used for the present-day climatology included the zonal and meridional 2-m wind speeds, which needed to be rotated on a per-grid-cell basis to get the x- and y-components of the wind vector. However, the GCM snapshots we used back then did not contain wind speed data, so I turned off that part of the code, implying that the wind field did not change during glacial cycles (a horrible assumption, really!).

TijnBerends reviewed (approved):
Much better!
